grails-app/controllers/org/pih/warehouse/api/StockMovementApiController.groovy

```diff
@@ -338,77 +336,30 @@ class StockMovementApiController {
-
-                // FIXME Should find bin location by name and parent and inventory item by lot number and expiration date
-                // and compare object equality (or at least PK equality) rather than comparing various components
-                AvailableItem availableItem = pickPageItem.availableItems?.find {
-                    (binLocation ? it.binLocation?.name == binLocation : !it.binLocation) && lotNumber == (it.inventoryItem?.lotNumber ?: null) &&
-                            expirationDate == (it?.inventoryItem?.expirationDate ? it.inventoryItem.expirationDate.format(Constants.EXPIRATION_DATE_FORMAT) : null)
-                }
+        MultipartFile importFile = command.importFile
+        if (importFile.isEmpty()) {
+            throw new IllegalArgumentException("File cannot be empty")
+        }
 
-                if (!availableItem) {
-                    throw new IllegalArgumentException("There is no item available with lot: ${lotNumber ?: ""}, expiration date: ${tokens[2] ?: ""} and bin: ${binLocation ?: ""}")
-                }
+        String csv = new String(importFile.bytes)
+        List<Map> importedLines = stockMovementService.parsePickCsvTemplateImport(csv)
 
-                RequisitionItem requisitionItem = pickPageItem.requisitionItem?.modificationItem ?: pickPageItem.requisitionItem
+        stockMovementService.validatePicklistListImport(importedLines, pickPageItems)
 
-                pickPageItem.picklistItems.each {
-                    if (it.id) {
-                        it.quantity = 0
-                    }
-                }
-                pickPageItem.picklistItems.add(new PicklistItem(
-                        requisitionItem: requisitionItem,
-                        inventoryItem: availableItem.inventoryItem,
-                        binLocation: availableItem.binLocation,
-                        quantity: quantityPicked,
-                        sortOrder: pickPageItem.sortOrder
-                ))
-            }
+        List errors = importedLines*.errors.withIndex().collect { errors, index ->
+            errors.collect { "Row ${1}: ${it}" }
+        }.flatten()
 
-            stockMovementService.createOrUpdatePicklistItem(stockMovement, pickPageItems)
+        stockMovementService.importPicklistTemplate(importedLines, stockMovement, pickPageItems)
 
-        } catch (Exception e) {
-            // FIXME The global error handler does not return JSON for multipart uploads
-            log.warn("Error occurred while importing CSV: " + e.message, e)
-            response.status = 500
-            render([errorCode: 500, errorMessage: e?.message ?: "An unknown error occurred during import"] as JSON)
-            return
+        if (!errors.isEmpty()) {
+            render([message: "Data imported with errors", errors: errors] as JSON)
         }
 
-        render([data: "Data will be imported successfully"] as JSON)
+        render([message: "Data imported successfully"] as JSON)
```

Comment:
I'm wondering why we don't return the imported items in the response

Reply:
This is the pattern that we have followed in most of the places where we update some data in the table and folow up with a GET request.

We could talk about it more and have a concrete pattern to follow for future implementations.

Reply:
I think it's safer to refretch the pickPageItems through the normal mechanism. That way the endpoint here can be free of a potential performance problem if the picklist is large. Let the client fetch deal with the usability problem of a slow load.